### PR TITLE
Add rapids_cpm_cccl feature.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR adds a new function `rapids_cpm_cccl` that adds CCCL from the new monorepo. This will eventually replace `rapids_cpm_thrust` and `rapids_cpm_libcudacxx`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-cmake/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The `cmake-format.json` is up to date with these changes.
- [ ] I have added new files under rapids-cmake/
   - [ ] I have added include guards (`include_guard(GLOBAL)`)
   - [ ] I have added the associated docs/ rst file and update the api.rst

### CODE REVIEW

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: rapids-cmake/cpm/cccl.cmake [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+Result Variables
+^^^^^^^^^^^^^^^^
+  :cmake:variable:`CCCL_SOURCE_DIR` is set to the path to the source directory of CCCL.
+  :cmake:variable:`CCCL_BINARY_DIR` is set to the path to the build directory of  CCCL.
+  :cmake:variable:`CCCL_ADDED`      is set to a true value if CCCL has not been added before.
+  :cmake:variable:`CCCL_VERSION`    is set to the version of CCCL specified by the versions.json.
+
+#]=======================================================================]
+# cmake-lint: disable=R0915
+function(rapids_cpm_cccl NAMESPACE namespaces_name)
```
[the remark this answers is not in the document]
Comment (bdice) [reply rapids-cmake/cpm/cccl.cmake:64]: I expected that we need to support a custom namespace argument that we propagate to Thrust / CUB somehow? Was the need for custom namespaces changed between 1.17.2 and 2.2.0? It looks like that custom namespace code is still around in Thrust/CUB.

https://github.com/NVIDIA/cccl/blob/v2.2.0/thrust/thrust/detail/config/namespace.h#L31

[the remark this answers is not in the document]
Comment (robertmaynard) [reply rapids-cmake/cpm/cccl.cmake:64]: The custom namespace that Thrust uses now is `CCCL`

File: rapids-cmake/cpm/cccl.cmake [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  set(multi_value)
+  cmake_parse_arguments(_RAPIDS "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  # TODO how to update this for CCCL?
+  set(post_find_code "if(NOT TARGET ${namespaces_name}::Thrust)"
+                     "  thrust_create_target(${namespaces_name}::Thrust FROM_OPTIONS)" "endif()")
+
+  if(Thrust_SOURCE_DIR)
+    # Store where CMake can find the Thrust-config.cmake that comes part of Thrust source code
+    include("${rapids-cmake-dir}/export/find_package_root.cmake")
```
[the remark this answers is not in the document]
Comment (bdice) [reply rapids-cmake/cpm/cccl.cmake:106]: I tried a few ways of updating this with the comments above/below but I can't figure out how to make the tests pass after changing this. Can you help me refactor this?

File: rapids-cmake/cpm/cccl.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdice) [reply rapids-cmake/cpm/cccl.cmake]: Okay. I gave this a shot. See if c8232bb45ee6fc25441d7a2c41545f1661fe5212 is what you expect.

[the remark this answers is not in the document]
Comment (bdice) [reply rapids-cmake/cpm/cccl.cmake]: How does removing the custom namespace affect cudf, which needs to have its own patched CCCL?

[the remark this answers is not in the document]
Comment (robertmaynard) [reply rapids-cmake/cpm/cccl.cmake]: It won't have an impact. Even currently we can only have a single version of Thrust or CCCL brought into a CMake project. The custom CMake namespace controls that thrust offered was due to an internal design choice that didn't need to be exposed to users/projects. Looking back on it rapids-cmake should have always just created something like a ` RAPIDS::Thrust` target

File: rapids-cmake/cpm/cccl.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertmaynard) [reply rapids-cmake/cpm/cccl.cmake:102]: Sure, I will at this today

[the remark this answers is not in the document]
Comment (robertmaynard) [reply rapids-cmake/cpm/cccl.cmake:102]: Looks to be a bug in Thrust/CCCL cmake logc that doesn't properly setup the targets on the first pass. Tracking down exactly what is wrong.

[the remark this answers is not in the document]
Comment (robertmaynard) [reply rapids-cmake/cpm/cccl.cmake:102]: https://github.com/NVIDIA/cccl/pull/1182

We can work around this for CMake 3.24+ users. For people on 3.23 the `CCCL::CCCL` target still works and will bring in thrust stuff as needed. I expect that most projects should migrate to `CCCL::CCCL` so this issue is minor and only exposed due to how our tests always verify all targets

[the remark this answers is not in the document]
Comment (robertmaynard) [reply rapids-cmake/cpm/cccl.cmake:105]: Pushed an update to this PR that works around the `CCCL::Thrust` issue, and removed unneeded logic from `cpm/cccl.cmake`.

File: rapids-cmake/cpm/versions.json [hunk 3]
```diff
@@ -5,11 +5,16 @@
...
+      "git_url" : "https://github.com/NVIDIA/cccl.git",
+      "git_tag" : "v${version}"
+    },
     "cuco" : {
       "version" : "0.0.1",
       "git_shallow" : false,
-      "git_url" : "https://github.com/NVIDIA/cuCollections.git",
-      "git_tag" : "7c76a124df0c2cd3fd66e3e080b9470a3b4707c6"
+      "git_url" : "https://github.com/bdice/cuCollections.git",
+      "git_tag" : "cccl-2.2.0"
```
Comment (bdice) [inline rapids-cmake/cpm/versions.json:17]: Reverting cuCollections changes so we can merge this.
```suggestion
      "git_url" : "https://github.com/NVIDIA/cuCollections.git",
      "git_tag" : "7c76a124df0c2cd3fd66e3e080b9470a3b4707c6"
```

**Pull request merged**